datajoint_plus/base.py

```diff
@@ -0,0 +1,832 @@
+                for n in [k for k in inds.keys() if k not in ['headers']]:
+                    if len(inds[n])>0:
+                        inds[n][inds[n] >= header_ind] += 1
+            
+            # reform and set definition
+            cls.definition = reform_definition(inds, contents)
+
+    @classmethod
+    def parse_hash_info_from_header(cls):
+        """
+        Parses hash_name and hashed_attrs from DataJoint table header and sets properties in class. 
+        """
+        header = cls.heading.table_info['comment']
+        matches = re.findall(r'\|(.*?);', header)
+        if matches:
+            for match in matches:
+                result = re.findall('\w+', match)
+
+                if result[0] == 'hash_name':
+                    cls.hash_name = result[1]
+
+                if result[0] == 'hashed_attrs':
+                    cls.hashed_attrs = result[1:]
+                
+                # handle booleans
+                for attr in ['hash_group', 'hash_table_name', 'hash_part_table_names']:
+                    if result[0] == attr:
+                        if result[1] == 'True' or result[1] == 'False':
+                            setattr(cls, attr, eval(result[1]))
+
+    @classmethod
+    def add_hash_to_rows(cls, rows, overwrite_rows=False):
+        """
+        Adds hash to rows. 
+        
+        Warning: rows must be able to be safely converted into a pandas dataframe.
+        :param rows (pd.DataFrame, QueryExpression, list, tuple): rows to pass to DataJoint `insert`.
+        :param hash_table_name (bool): Whether to include table_name in rows for hashing
+        :overwrite_rows (bool): Whether to overwrite key/ values in rows. If False, conflicting keys will raise a ValidationError. 
+        :returns: modified rows
+        """
+        assert cls.hashed_attrs is not None, 'Table must have hashed_attrs defined. Check if hashing was enabled for this table.'
+
+        hash_table_name = True if cls.hash_table_name or (issubclass(cls, dj.Part) and hasattr(cls.master, 'hash_part_table_names') and getattr(cls.master, 'hash_part_table_names')) else False
+
+        if hash_table_name:
+            table_name = {'#__table_name__': cls.table_name}
+        else:
+            table_name = None
+            
+        rows = format_rows_to_df(rows)
+
+        for a in cls.hashed_attrs:
+            assert a in rows.columns.values, f'hashed_attr "{a}" not in rows. Row names are: {rows.columns.values}'
+
+        if _is_overwrite_validated(cls.hash_name, rows, overwrite_rows):
+            rows_to_hash = rows[[*cls.hashed_attrs]]
+
+            if cls.hash_group:
+                rows[cls.hash_name] = generate_hash(rows_to_hash, add_constant_columns=table_name)[:cls.hash_len]
+
+            else:
+                rows[cls.hash_name] = [generate_hash([row], add_constant_columns=table_name)[:cls.hash_len] for row in rows_to_hash.to_dict(orient='records')]
```

Comment:
When a `hash_part_table` has no secondary attributes, this line breaks since `rows_to_hash.to_dict(orient='records')` will return `[]` instead of `[{}]`. 
`hash_part_table`s with no secondary attributes can be useful for semantic notations for methods with no parameters, for example:
```
class AreaMemberShip(djp.Manual):  # table storing a mapping from functional unit keys to brain area
    definition = """
    -> fuse.ScanSet.Unit
    ---
    brain_area
    """


class AreaConfig(djp.Lookup):  # method table logging different ways of mapping unit_keys to brain areas
    class AreaMemberShip(djp.Part):  # load brain area from the above table AreaMemberShip, this table does not need any secondary keys
        definition = """
        -> master
        ---
        """
```

Reply:
I will turn this into an issue once the PR is merged.